## Welcome to GitHub Pages

You can use the [editor on GitHub](https://github.com/volpatoo/volpatoo.github.io/edit/main/index.md) to maintain and preview the content for your website in Markdown files.

Whenever you commit to this repository, GitHub Pages will run [Jekyll](https://jekyllrb.com/) to rebuild the pages in your site, from the content in your Markdown files.

### Markdown

Markdown is a lightweight and easy-to-use syntax for styling your writing. It includes conventions for

```markdown
Syntax highlighted code block

# Header 1
## Header 2
### Header 3

- Bulleted
- List

1. Numbered
2. List

**Bold** and _Italic_ and `Code` text

[Link](url) and ![Image](src)
```

For more details see [Basic writing and formatting syntax](https://docs.github.com/en/github/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax).

### Jekyll Themes

Your Pages site will use the layout and styles from the Jekyll theme you have selected in your [repository settings](https://github.com/volpatoo/volpatoo.github.io/settings/pages). The name of this theme is saved in the Jekyll `_config.yml` configuration file.

### Support or Contact

Having trouble with Pages? Check out our [documentation](https://docs.github.com/categories/github-pages-basics/) or [contact support](https://support.github.com/contact) and weâ€™ll help you sort it out.



---
title: "Genomic Selection for White Mold GS Project NSI"
author: "Leonardo Volpato"
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    keep_md: yes
    toc: false
    number_sections: true
fig_caption: yes
link-citations: true
classoption: a4paper
---

```{r global_options, include = FALSE}
knitr::opts_chunk$set(comment = "#>", 
                      collapse = TRUE,
                      echo = TRUE,
                    	message = FALSE,
                    	warning = FALSE
                      #fig.dim = c(7, 4)
)
op <- options(width = 90)
options(tinytex.verbose = TRUE)
```

# About this analysis {-}

The present analysis aims to dissect the genomic selection (GS) data using from the white mold Dry Beans breeding project - MSU. 

The trait in the study is the ratings (1-9) measurements adjusted to the BLUEs values and area per meter square. 

The following methods of analysis will be done to lead with the GS performance:

* [Input phenotype data set](#phenodata)
* [Preparing the pheno data to be used for genomic prediction](#preddata)
* [Input genotypic data set and adjusting the martkers values](#datamartk)
* [Select markers and calculate marker effects](#mrkseff)
* [Implementing the G-BLUP analysis](#gblup)
  * [Black Beans G-BLUP analysis](#gblupBB)
  * [Navy Beans G-BLUP analysis](#gblupNB)
  * [Red Pink Beans G-BLUP analysis](#gblupRB)
* [CROSS VALIDATION n-fold](#eaMktCV)
  * [Black and Navys beans market class combined](#CVNAandBB)   
* [LEAVE-ONE-OUT CROSS VALIDATION LOOCV](#LOOCVeaMkt)
  * [Black and Navys beans market class combined](#LOOCVNAandBB)
* [GENETIC ALGORITHM FOR SUBSET SELECTION](#STPGA)
  * [Black and Navys beans market class combined](#STPGANAandBB)


---


## Setting up the working directory {-} 
```{r directory}
rm(list=ls())
my.path <- dirname(rstudioapi::getActiveDocumentContext()$path)
setwd(my.path)
# R setup ###########
options("scipen"=999,"digits" =5) # set numbering format
```

## Load and install necessary Packages {-}
```{r loading r packages, echo=FALSE}
if(!require("pacman")) install.packages("pacman")

pacman::p_load(rrBLUP,
               dplyr,
               tidyverse,
               ggplot2,
               broom,
               STPGA,
               RColorBrewer,
               BGLR,
               ggpubr
               
)
```

# Input phenotype data set {#phenodata}
The data set used to perform this analysis was generated by Dr. Francisco Gomez, MSU 2022.
```{r Input phenotype data set}
pheno2<-read.csv(file="pheno.csv", header = TRUE)
colnames(pheno2)
```

# Preparing the pheno data to be used for genomic prediction {#preddata}
## Yield trait (Kg/ha)
* The  BLUE_pct_wm2 s and BLUPs yield has a high correlation.\n
* The BLUE_pct_wm2 s for yield data will be used to perform all analysis, in which this values comes from a simple model Yij = u + Gi + Bj + Eij.\n
* The yield was corrected from Hundredweights Per Acre (Cwt/ac) to Kilograms Per Hectare (kg/ha) and the trait name was adjusted.
* Subset MSU lines only.
* NA removed.
* Black, Nayvs and Red Pink pheno data subseted 


```{r data preparation}
pheno <- pheno2[,c("Genotype", "Class", "Type", "BLUE_cwt_a","BLUE_pct_wm2","BLUE_wm")]
cortraits<- cor.test(pheno2$BLUE_pct_wm2, pheno2$BLUE_wm)
cortraits

phenoMSU <- subset(pheno, Type=="MSU")
phenoNA <-na.omit(phenoMSU)
dim(phenoMSU)

#Subset Black pheno
phenoB <- subset(phenoNA, Class=="Black")

#Subset Navy pheno
phenoN <- subset(phenoNA, Class=="Navy")

#Subset Red pheno
phenoR <- subset(phenoNA, Class=="Red_Pink")

```

Boxplot of the original data set to Yield (KG/HA)
```{r boxplot pheno data, fig.width=9, fig.height=4, fig.cap="BoxPlot showing the BLUE_pct_wm2  (x) distribuation by class (y)"}

plot1<- phenoNA %>% ggboxplot( y="BLUE_cwt_a", x = "Class",
                    fill="gray", ylab ="Observed White molde",
                    bxp.errorbar = T) + coord_flip()
plot1

plot2<-phenoNA %>% ggboxplot( y="BLUE_wm", x = "Class",
                    fill="gray", ylab ="Observed White molde",
                    bxp.errorbar = T) + coord_flip()
plot2

plot3<-phenoNA %>% ggboxplot( y="BLUE_pct_wm2", x = "Class",
                    fill="gray", ylab ="Observed White molde",
                    bxp.errorbar = T) + coord_flip()
plot3


```


# Input genotypic data set and adjusting the martkers values{#datamartk}
* Black, Nayvs and Red Pink geno data subseted 

```{r genotypic info}
geno<-read.csv(file = "geno2.csv", row.names = 1)
geno <-data.matrix(geno) #Convert geno data to numeric
is.numeric(geno) #check if it is numeric
# replacing the markers values.
geno1<-geno #Keeping the original file
geno1[which(geno1=="0")] <- -1 
geno1[which(geno1=="0.5")] <- 0
geno1[which(geno1=="1")] <- 1

geno1<-as.matrix(geno1) # classe do objeto como matriz
dim(geno1) #192 3889


#Subset Black geno
# Subset lines with genotype that were only subssted for the phenotype training set "SUBSET"
genoB <- subset(geno1, row.names(geno) %in% phenoB$Genotype)

#Subset Navy geno
genoN <- subset(geno1, row.names(geno) %in% phenoN$Genotype)

#Subset Red geno
genoR <-subset(geno1, row.names(geno) %in%phenoR$Genotype)


```
## Heatmap G relatioship matrix for the whole data set
```{r all heatmap, message=FALSE,warning=FALSE, fig.width=9, fig.height=8, fig.cap="Heatmap G relatioship matrix for the whole data set"}
#Used to compute the heatmap G relatioship matrix
Gm_total<-A.mat(X = geno1, min.MAF = 0.05)
scaleorange <- colorRampPalette(brewer.pal(9, "Oranges"))(9)
G.heatmap<- heatmap(Gm_total, labRow=NULL, labCol=NA, col=scaleorange, cexRow=NULL)
```

# Select markers and calculate marker effects {#mrkseff}

* Use model selection algorithm described in the lecture to select markers and calculate marker effects
* The entire data set `pheno` and `geno` were used versus trait yield (`BLUE_pct_wm2 `)
```{r mrkseff model}
##Source in function for model selection
source('modelSelFunc.r')

##Load provided data. Marker data is coded as -1, 0, 1, where -1 is arbitrarily assigned to the allele coming from parent one, 0 is assigned to heterozygotes, and 1 is assigned to the allele coming from parent two.
pheno.1 <- pheno
mrks <- geno1

##Use model selection algorithm described in the lecture to select markers and calculate marker effects.
pVec <- vector(length=dim(mrks)[2])
dim(mrks)
dim(pheno.1)
for (i in 1:dim(mrks)[2]){
	pVec[i] <- summary(lm(pheno.1$BLUE_pct_wm2  ~ mrks[, i]))[[4]][2, 4]
	}
 
sigMrk <- which(pVec < 0.05)
#length(sigMrk)
mrks2 <- as.matrix(mrks[, sigMrk]) 
#dim(mrks2)
lsModel <- selectModel(pheno.1$BLUE_pct_wm2 , q1=matrix(1, nrow=dim(mrks2)[1], ncol=1), geno=as.data.frame(mrks2), p.in = 0.05, p.out = 0.05, verbose = F, max.rep = 1000)

##Pull out selected markers
ndx <- match(names(lsModel$coefficients)[-1], colnames(mrks))

mrkEffsOLS <- lsModel$coefficients[-1]


##Now estimate marker effects using rrBLUP package and make predictions in test set
rrModel <- mixed.solve(y=pheno.1$BLUE_pct_wm2 , Z=mrks)

mrkEffsRR <- rrModel$u

##Use marker effects to calculate genomic estimated breeding values of individuals in training set. Here we are extracting the intercept and adding it back on.
int <- as.numeric(rrModel$beta)
rrGebv <- int + mrks%*%mrkEffsRR

t<- data.frame(c(mrkEffsRR, mrkEffsOLS))
length(mrkEffsOLS)
```


```{r mrkseff model plot, fig.width=9, fig.height=5, fig.cap="Plot marker effects"}
##Plot marker effects where x-axis is just marker number in order and y-axis is marker effect. Black points are RR-BLUP marker effect predictions, and red points are least-squares estatimation.
plot(c(seq(1:3889), ndx), c(mrkEffsRR, mrkEffsOLS), ylab='mrk eff', xlab='Mrk num') 
points(ndx, mrkEffsOLS, col='red', pch=19)

```


# Implementing the G-BLUP analysis {#gblup}
**The rrBLUP package used to perform the analysis**

## *Black beans only G-BLUP*{#gblupBB}
**Genomic additive relationship matrix**
```{r a.mat BB}
##First, use rrBLUP package to calculate genomic additive relationship matrix
Gm<-A.mat(X = genoB, min.MAF = 0.05)
dim(Gm)
```
## Heatmap G relatioship matrix for Black Beans
```{r black heatmap,message=FALSE,warning=FALSE, fig.width=9, fig.height=8, fig.cap="Heatmap G relatioship matrix for Black Beans"}
#Used to compute the heatmap G relatioship matrix
scaleorange <- colorRampPalette(brewer.pal(9, "Oranges"))(9)
G.heatmap<- heatmap(Gm, labRow=NULL, labCol=NA, col=scaleorange, cexRow=NULL)
```

## Model sintase 
### `kin.blup` function use

**Model using the function `kin.blup` from the `rrBlup` package.**
```{r model BB}
set.seed(12345)
result <- kin.blup(data=phenoB, geno='Genotype', pheno='BLUE_pct_wm2', K=Gm)
# Genotipic variance
Va<-result$Vg
Va
# Residual variance
Ve<-result$Ve
Ve
#Herdability
h2a<- Va/(Va+Ve)
h2a

# Predicted values
y_hat<- result$pred

#dim(y_hat)
y_hat1 <- data.frame(y_hat)
y_hat1 <- y_hat1 %>% rownames_to_column(var = "Genotype")
y_hat1 <- arrange(y_hat1, Genotype)
pheno.add<-merge(phenoB, y_hat1, by = "Genotype")

# Prediction ability 

corr.kin.blup <- pheno.add  %$%
  cor.test(BLUE_pct_wm2 , y_hat) %>%
  tidy %>%
  mutate_if(is.numeric, round, 4)

corr.kin.blup

# Bies of prediction
pheno.add<-pheno.add[order(pheno.add$BLUE_pct_wm2 , decreasing = FALSE), ]

#lm(formula = pheno.add$BLUE_pct_wm2  ~ pheno.add$y_hat)

bies.kin.blup<- 1 - lm(formula = pheno.add$BLUE_pct_wm2  ~ pheno.add$y_hat)[[1]][2]
  bies.kin.blup<- bies.kin.blup %>%
    as_tibble()
bies.kin.blup
```

 


